Colab_DAIN.ipynb

```diff
@@ -171,98 +158,58 @@
     {
       "cell_type": "code",
       "metadata": {
-        "id": "XunYRBtyPZf_",
-        "colab_type": "code",
-        "colab": {}
+        "id": "e5AHGetTRacZ",
+        "cellView": "form"
       },
       "source": [
-        "# Install known used versions of PyTorch and SciPy\n",
-        "!pip install torch==1.4.0+cu100 torchvision==0.5.0+cu100 -f https://download.pytorch.org/whl/torch_stable.html\n",
-        "!pip install scipy==1.1.0\n",
+        "#@title Setup everything. This takes a while. Just wait ~20 minutes in total.\n",
+        "\n",
+        "# Install old pytorch to avoid faulty output\n",
+        "%cd /content/\n",
+        "!wget -c https://repo.anaconda.com/miniconda/Miniconda3-4.5.4-Linux-x86_64.sh\n",
+        "!chmod +x Miniconda3-4.5.4-Linux-x86_64.sh\n",
+        "!bash ./Miniconda3-4.5.4-Linux-x86_64.sh -b -f -p /usr/local\n",
+        "!conda install pytorch==1.1 cudatoolkit torchvision -c pytorch -y\n",
+        "!conda install ipykernel -y\n",
```

Comment:
Interesting choice. Are there advantages to installing pytorch through conda instead of pip?

Reply:
I personally prefer conda over pip for this task, but you could use both. Installing pytorch with conda works more easy and reliable than pip in my experience, since conda does configure the needed environment automatically. Aside that, cudatoolkit is independent from the installed cuda, so it won't make problems if Colab changes the local Cuda version.

Installing older versions with pip usually tends to require the usage of .whl files due to missing versions with "==". You solved this by giving pip a link. Conda does not show such a limitation and has a big list of available versions, which you can find here: https://anaconda.org/pytorch/pytorch/files?page=1

The last reason would be, that I used this code in the past and I know it works. :P

Reply:
I see, thanks for the explanation. It absolutely makes sense. The programmer in me is scared when adding new dependencies to a project but your explanation absolutely justifies your choice. Thank you!

Reply:
One small update: It is true that using conda will make the enviroment independent from system python and make it robust against python updates within colab, but [pytorch is bundled with cuda and does not require cuda externally](https://github.com/pytorch/pytorch/issues/47109#issuecomment-745835415). The package cudatoolkit will be automatically downloaded with conda nonetheless for some reason and that's why I assumed it is certainly needed. 

```
conda install -c pytorch pytorch
    package                    |            build
    ---------------------------|-----------------
    cudatoolkit-11.0.221       |       h74a9793_0       627.0 MB
    pytorch-1.7.1              |       py3.8_cuda110_cudnn8_0      1007.0 MB  pytorch
```

My other points are still valid. One random point I could add would be download speed. Not sure which one would be faster.